quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,282 @@
+    }
+    fn exists(&self, path: &Path) -> bool {
+        self.files.contains_key(path)
+    }
+}
+
+#[async_trait]
+impl Storage for BundleStorage {
+    async fn put(&self, path: &Path, _payload: crate::PutPayload) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn copy_to_file(&self, path: &Path, _output_path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> crate::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
+        Ok(self.metadata.exists(path))
+    }
+
+    fn uri(&self) -> String {
+        self.storage.uri()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+struct FileOffsets {
+    start: u64,
+    end: u64,
+}
+
+impl HasLen for BundleStorage {
+    fn len(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+impl fmt::Debug for BundleStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BundleStorage({:?}, files={:?})",
+            &self.bundle_file_name, self.metadata
+        )
+    }
+}
+
+fn unsupported_operation(path: &Path) -> StorageError {
+    let msg = "Unsupported operation. BundleStorage only supports async reads";
+    error!(path=?path, msg);
+    io::Error::new(io::ErrorKind::Other, format!("{}: {:?}", msg, path)).into()
+}
+
+/// BundleStorage bundles together multiple files into a single file
+/// with some metadata
+pub struct BundleStorageBuilder {
+    metadata: BundleStorageMetadata,
+    current_offset: u64,
+    /// The offset in the file where the hotcache begins. This is used to read
+    /// the hotcache and footer in a single read.
+    pub hotcache_offset: u64,
+    /// The footer offset, where the footer begins.
+    /// The footer includes the footer metadata as json encoded and the size of the footer in
+    /// bytes.
+    pub footer_offset: u64,
+    bundle_file: File,
+}
+
+impl BundleStorageBuilder {
+    /// Creates a new CreateBundleStorage, to which files can be added.
+    pub fn new(path: &Path) -> io::Result<Self> {
```

Comment:
Taking an io::Write would be nicer.

If you want to get rid of the generic, you can just make new generic, wrap it in a BufWriter<Box<dyn io::Write>>

Reply:
I would keep `Path` for now, since we create the bundle in the local fs and then upload it, the creation and ownership can be in the builder

Reply:
That's a bottom-up vs top-bottom debate.
It's tricky because there are Pros and Cons to both.

The cons of bottom-up are mild here:
- We don't really hide the intent too much.
- We don't introduce complicated abstraction that impact the complexity.

The pros are:
- The code reuse is easier. I agree we need to consider that one mild.
We are not closing any doors by using Path.
- Unit tests will be cleaner and won't write to disk.

Reply:
I let you decide.